pip_audit/_dependency_source/requirement.py

```diff
@@ -225,18 +225,23 @@ def _fix_file(self, filename: Path, fix_version: ResolvedFixVersion) -> None:
         # Check ahead of time for anything invalid in the requirements file since we don't want to
         # encounter this while writing out the file. Check for duplicate requirements and lines that
         # failed to parse.
-        req_names: set[str] = set()
+        req_specifiers: dict[str, SpecifierSet] = dict()
+
         for req in reqs:
             if (
                 isinstance(req, InstallRequirement)
                 and (req.marker is None or req.marker.evaluate())
                 and req.req is not None
             ):
-                if req.name in req_names:
+                duplicate_req_specifier = req_specifiers.get(req.name)
+
+                if not duplicate_req_specifier:
+                    req_specifiers[req.name] = req.specifier
+
+                elif duplicate_req_specifier != req.specifier:
```

Comment:
Just making sure I understand the behavior here: rather than deduping on just the requirement's `name` (i.e. package name), we're now deduping on the conflicting specifiers for duplicated names, right?

IIUC that means that we'll still have some false positives here (since `SpecifierSet != SpecifierSet` does not imply that the two are incompatible), right? If so that's fine, just making sure I understand 🙂

Reply:
> rather than deduping on just the requirement's name (i.e. package name), we're now deduping on different specifiers for duplicated names

Yep, that's the idea. As long as each package sharing the same name, also share the same specifier, it is considered identical.

> IIUC that means that we'll still have some false positives here (since SpecifierSet != SpecifierSet does not imply that the two are incompatible), right? If so that's fine, just making sure I understand 🙂

I'm not sure how we could still get false positives here 🤔 Have you a specific case in mind ? 
Unless you modify the resolved requirements, it is very unlikely to get different `SpecifierSet` for the package and its extras.

Reply:
> I'm not sure how we could still get false positives here 🤔 Have you a specific case in mind ?
Unless you modify the resolved requirements, it is very unlikely to get different SpecifierSet for the package and its extras.

What I'm thinking of is something like this:

```
example >= 1.0
example >= 1.0, < 3.0
```

In this case, `SpecifierSet(">= 1.0") != SpecifierSet(">= 1.0, < 3.0")`, but the two are not incompatible (all versions in 1.x and 2.x satisfy both sets). Unless I'm misunderstanding, I believe this change will still incorrectly reject this as a "duplicate" dependency, when one is a superset of the other.

Reply:
You are correct, it this case it will be rejected.
Maybe it would be more appropriate to reject them as `incompatible` instead of `duplicate`, what do you think ?

Reply:
> Maybe it would be more appropriate to reject them as incompatible instead of duplicate, what do you think ?

Actually, I think I misunderstood the context here, sorry! From a quick look I think we can never reach this kind of double-range state, since `--disable-pip` means that we always have a fully-pinned requirements file.

So I think we can continue to describe this as `duplicate`.